File: job-offer/src/main/java/com/oxygensend/joboffer/infrastructure/fixtures/AddressFixture.java
package com.oxygensend.joboffer.infrastructure.fixtures;

import com.oxygensend.joboffer.domain.entity.Address;
import com.oxygensend.joboffer.domain.repository.AddressRepository;
import com.oxygensend.joboffer.infrastructure.services.staticdata.dto.AddressDto;
import com.oxygensend.springfixtures.Fixture;
import com.oxygensend.springfixtures.FixtureType;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.web.client.RestClient;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class AddressFixture implements Fixture {
    public final static int SIZE = 3000;
    private final AddressRepository addressRepository;
    private final RestClient restClient;
    private final String staticDataUrl;

    AddressFixture(AddressRepository addressRepository, RestClient restClient, String staticDataUrl) {
        this.addressRepository = addressRepository;
        this.restClient = restClient;
        this.staticDataUrl = staticDataUrl;
    }

    @Override
    public void load() {
        List<Address> addresses = new ArrayList<>();
        List<AddressDto> importedAddresses = restClient.get()
                                                       .uri(staticDataUrl
                                                                + "/api/v1/static-data/addresses/with-postal-codes")
                                                       .retrieve()
                                                       .body(new ParameterizedTypeReference<>() {
                                                       });
        Collections.shuffle(importedAddresses);

        for (int i = 0; i < SIZE; i++) {
            var address = importedAddresses.get(i);
            addresses.add(new Address(address.city(), address.postalCodes().getFirst(), address.lat(), address.lon()));
        }

        addressRepository.saveAll(addresses);
    }


    @Override
    public boolean isEnabled() {
        return addressRepository.count() == 0;
    }

    @Override
    public String collectionName() {
        return "addresses";
    }

    @Override
    public FixtureType type() {
        return FixtureType.DICTIONARY;
    }
}

File: static-data/src/main/kotlin/com/oxygensend/staticdata/infrastructure/fixtures/AddressFixture.kt
package com.oxygensend.staticdata.infrastructure.fixtures

import com.github.javafaker.Faker
import com.oxygensend.springfixtures.Fixture
import com.oxygensend.springfixtures.FixturesFakerProvider
import com.oxygensend.staticdata.domain.Address
import com.oxygensend.staticdata.domain.AddressRepository
import org.bson.types.ObjectId
import org.springframework.stereotype.Component
import java.util.*
import java.util.stream.Stream

@Component
class AddressFixture(
    private val addressRepository: AddressRepository,
    fakerProvider: FixturesFakerProvider
) : Fixture {

    private val random: Random = fakerProvider.random()
    private val faker: Faker = fakerProvider.faker()

    override fun load() {
        val addresses: List<Address> = Stream.generate {
            val address = Address(ObjectId().toHexString())
            address.city = faker.address().cityName()
            address.lat = faker.address().latitude().toDouble()
            address.lon = faker.address().longitude().toDouble()
            address.postalCodes = setOf(faker.address().zipCode())
            return@generate address
        }
            .distinct()
            .limit(500)
            .toList()

        addressRepository.saveAll(addresses);
    }

    override fun isEnabled(): Boolean = addressRepository.count() == 0.toLong()

    override fun collectionName(): String {
        return "addresses"
    }
}